zephyr/lib/alloc.c

```diff
@@ -163,7 +181,13 @@ static void *heap_alloc_aligned(struct k_heap *h, size_t min_align, size_t bytes
 #endif
 
 	key = k_spin_lock(&h->lock);
+#if CONFIG_L3_HEAP
+	l3_heap_invalidate(h);
+#endif
```

Comment:
this looks strange... You just took a spinlock at `&h->lock` which presumably will modify RAM contents, i.e. its cache, and now you invalidate caches. Is that correct? Why is this invalidation needed at all?

Reply:
The k_heap structure and its spinlock are located in HPSRAM, while l3_heap_invalidate() invalidates cache entries for L3 address range. There is z_heap zephyr internal structure that must be updated each time new alloc/free operation is done. Therefore previous implementation based on uncashed addresses and only pointers to memory buffers itself were translated to cached address. However for L3 Memory on ACE the recommendation is to use always cached addresses.